### ISSUE TITLE
Graphical glitches [python 3.10?]

### ISSUE THREAD

**Issue by rien333** [qutebrowser/qutebrowser#6890]
See this [reddit thread](https://www.reddit.com/r/qutebrowser/comments/rftxk8/qutebrowser_v240_window_flickering_random/).

**Version info**:

```
qutebrowser v2.4.0
Git commit: 297b33928 on master (2021-12-12 11:12:16 +0100)
Backend: QtWebEngine 5.15.7, based on Chromium 87.0.4280.144
Qt: 5.15.2

CPython: 3.10.1
PyQt: 5.15.6

sip: 6.1.0.dev2104271705
colorama: 0.4.4
jinja2: 3.0.3
pygments: 2.10.0
yaml: 5.4.1
adblock: 0.5.1
PyQt5.QtWebEngineWidgets: yes
PyQt5.QtWebEngine: 5.15.5
PyQt5.QtWebKitWidgets: no
pdf.js: no
sqlite: 3.37.0
QtNetwork SSL: OpenSSL 1.1.1l  24 Aug 2021

Style: QFusionStyle
Platform plugin: xcb
OpenGL: Intel, 4.6 (Compatibility Profile) Mesa 21.3.1
Platform: Linux-5.15.7-arch1-1-x86_64-with-glibc2.33, 64bit
Linux distribution: Arch Linux (arch)
Frozen: False
Imported from /usr/lib/python3.10/site-packages/qutebrowser
Using Python from /usr/bin/python3
Qt library executable path: /usr/lib/qt/libexec, data path: /usr/share/qt

Paths:
cache: /home/rw/.cache/qutebrowser
config: /home/rw/.config/qutebrowser
data: /home/rw/.local/share/qutebrowser
runtime: /run/user/1000/qutebrowser
system data: /usr/share/qutebrowser
```

**Does the bug happen if you start with `--temp-basedir`?**:

Interestingly enough, no, it does not happen with `-T`. Perhaps something cache related? Someone over at reddit said it wasn't config related.

**Description**

Qutebrowser started exhibiting all sorts of weird graphical glitches at a moment everyone on reddit (myself included) ascribes to the release of python3.10 (on arch linux). For me, just switching tabs is enough for the glitchiness to show. See the video below, and also the video linked in the thread above. 

https://user-images.githubusercontent.com/10417027/145956926-6413eaae-63d4-408b-9702-eab538170404.mp4

**How to reproduce**

Not sure yet?

**Comment by rien333** [comment]
I can not reproduce the issue anymore when I remove `enable-gpu-rasterization` from `qt.args`. Apparently a lot of people have this flag set? Feel free to close, especially if anyone else can confirm.

**Comment by toofar** [comment]
There was this issue about display issues the other day, no idea if it is related but it's not like these reports come up often so it might be! https://github.com/qutebrowser/qutebrowser/issues/6879

**Comment by rien333** [comment]
Interesting, let me check if 1) mesa recently upgraded on arch, and whether downgrading changes stuff.

**Comment by The-Compiler** [comment]
It's definitely not Python 3.10 (we've been testing that ever since the first Alpha in January). Agreed it's pretty likely to be mesa.

**Comment by steveb-123** [comment]
I get this with `-T` temp basedir also. Perhaps it its less frequent, though I'm not sure. It's still unusable on many sites sadly.

I'm on arch, latest everything. Installed via `community` or `-git` packages

Non-scientifically, the following things seem to make it worse:

* Complex sites: 
   * arch wiki is okay; 
   * sites with lots of 'frames', ads, video boxes = bad; 
   * github = in between
* Increasing the number of screen pixels being rendered (?) by the browser. That means adjusting window to most of screen width and zooming content so it occupies the full ~3840 width on my display
* Scrolling up and down in one of these 'complex' sites
* *Maybe*, using my default config rather than `-T`

For me, these things do *not* fix the problem:

* `stop` ing the page
* moving around the page with a different input device (eg touch screen, mouse, kb)
* Switching between git and community packages on arch
* reinstalling all the optional dependencies listed in arch package
* reinstalling every part of `sway`, `wayland`, `xorg` etc
* running `sway` with hidpi patches or not.
* running `qutebrowser` +/- .Xresources and .Xdefaults
* installing (or not) the `qt5-wayland` package in arch

**Comment by The-Compiler** [comment]
Well, the most interesting and also most likely open question is still:

> let me check if 1) mesa recently upgraded on arch, and whether downgrading changes stuff

**Comment by steveb-123** [comment]
I'm attemping to switch to `mesa-git` right now, though the clone is coming in very slowly. I'll report back if that fixes.

**Comment by The-Compiler** [comment]
Note the bug report over at mesa is still open, so I don't think that will help - worth a try anyways I guess.

**Comment by rien333** [comment]
Yes, it seems to be mesa-related. Downgrading seems to have solved the problem, also perhaps because mesa was recently updated on arch?

For all arch users:

Either use `downgrade` from the AUR to downgrade mesa, or run something like `sudo pacman -U /var/cache/pacman/pkg/mesa-21.2.5-1-x86_64.pkg.tar.zst`.

**Comment by steveb-123** [comment]
fyi for anyone wondering, switching to `mesa-git`, this morning anyhow.

**Comment by rien333** [comment]
Does downgrading to an older version help?

**Comment by steveb-123** [comment]
Yes, it did. It took me a minute as I just needed to confirm.

I only had a couple of versions on my system and I misread the numbers. For me, on arch, the following switch sorted things right out:

`sudo pacman -U /var/cache/pacman/pkg/mesa-21.2.5-1-x86_64.pkg.tar.zst`

Would this be worth flagging with the arch packagers in some way?

**Comment by rien333** [comment]
> Would this be worth flagging with the arch packagers in some way?

I don't think they will be keen to roll back a whole package release, especially because it doesn't affect all of the userbase, but since the bug is quite significant, they might be open to a patch as soon as the mesa team comes up with a fix. Alternatively, if the mesa team recognizes this bug as significant, they might be open to a quick bug-fix release.

**Comment by markstos** [comment]
Here's more detailed rollback plan for Arch:

```
sudo pacman -U /var/cache/pacman/pkg/*mesa*21.2.5*.zst
```
Then edit `/etc/pacman.conf` and add the packages to `IgnorePkgs`:

```
IgnorePkg   = mesa, lib32-mesa
```

Now the packages won't be upgraded until you remove them from this file.

**Comment by rien333** [comment]
Thanks! `downgrade` from the aur does all that automatically FWIW, can't recommend it enough to the  lazy/forgetful computer user.

**Comment by wgmayer0** [comment]
> Downgr

I can confirm that downgrading fixed my problem. Thank you!

Does this mean that mesa has to fix this in their package? I am not entirely sure how open source works as far as who is responsible for what when it comes to fixing bugs. I would assume there is nothing Qutebrowser can do?

**Comment by The-Compiler** [comment]
Exactly - it's a bug in mesa, and it doesn't look like there is much qutebrowser can do about it.

**Comment by prosoitos** [comment]
For info, I opened [an issue on the Mesa repo](https://gitlab.freedesktop.org/mesa/mesa/-/issues/5745#note_1192463) yesterday.

**Comment by arxel-sc** [comment]
Hello. As another workaround, setting the environment variable "MESA_LOADER_DRIVER_OVERRIDE=llvmpipe" seems to work fine without downgrading the package.

**Comment by prosoitos** [comment]
@arxel-sc. I just gave it a try and it helps (no blanking or flashing) but fancy websites (e.g. https://www.nytimes.com/interactive/2021/12/13/climate/antarctic-climate-change.html) don't get rendered properly (at least for me).

So I'll downgrade again.

**Comment by The-Compiler** [comment]
Could someone who is affected by this (or someone who doesn't mind upgrading again) please try whether `QTWEBENGINE_DISABLE_GPU_THREAD=1` changes anything (should force webengine to single-threaded opengl mode, but is QtWebEngine 5.15 only)?

Personally I can't seem to reproduce this, but I'm on AMD graphics.

**Comment by moorchegue** [comment]
I can try. Is that an env var?

**Comment by qvtqht** [comment]
I just installed a fresh Manjaro, and I'm unsure how to downgrade to the correct Mesa version, since everything I've read online assumes that I already had the previous version installed, and it's in my package cache.

If anyone knows Manjaro/Arch well enough to post a solution for this, I'd really appreciate it. TIA!

**Comment by rien333** [comment]
@moorchegue 

> Is that an env var?

Yes.

@qvtqht 

Third time I've recommended it, but `downgrade` from the aur can fetch old versions from online, rather than local, sources.

**Comment by wgmayer0** [comment]
> Hello. As another workaround, setting the environment variable "MESA_LOADER_DRIVER_OVERRIDE=llvmpipe" seems to work fine without downgrading the package.

This gets me closer to the intended behavior but I still see plenty of flickering, etc. This is not an adequate solution in my situation so I will go back to downgrading.

**Comment by wgmayer0** [comment]
> Either use `downgrade` from the AUR to downgrade mesa, or run something like `sudo pacman -U /var/cache/pacman/pkg/mesa-21.2.5-1-x86_64.pkg.tar.zst`.

I have the ignore line in my config but it is still upgrading the package... hmm

**Comment by wgmayer0** [comment]
> Here's more detailed rollback plan for Arch:
> 
> ```
> sudo pacman -U /var/cache/pacman/pkg/*mesa*21.2.5*.zst
> ```
> 
> Then edit `/etc/pacman.conf` and add the packages to `IgnorePkgs`:
> 
> ```
> IgnorePkg   = mesa, lib32-mesa
> ```
> 
> Now the packages won't be upgraded until you remove them from this file.

Thanks for the more detailed plan. One edit is to separate the packages using only a space - if you use a comma, pacman will still upgrade the package.

See https://bbs.archlinux.org/viewtopic.php?id=237671

**Comment by prosoitos** [comment]
@qvtqht You can find all package versions [in the archive](https://wiki.archlinux.org/title/Arch_Linux_Archive)

**Comment by qvtqht** [comment]
@wgmayer0 Thank you very much for the detailed instructions! I was finally able to downgrade.

**Comment by qvtqht** [comment]
Unfortunately, after confirming that my lib32-mesa and mesa are at 21.2.5, and rebooting for good measure, I'm still experiencing quite a bit of glitching.

In particular, the tab bar flickers whenever any update happens. (tabs.position=right)

**Comment by LamprosPitsillos** [comment]
Can you check again?Mesa was updated to `mesa 21.3.2-1` , alot of issues where closed.

**Comment by kflak** [comment]
Nope, still glitching like crazy...

**Comment by kflak** [comment]
Nope, still glitching like crazy with 21.3.2-1, unfortunately...

**Comment by qvtqht** [comment]
@rien333

> Third time I've recommended it, but `downgrade` from the aur can fetch old versions from online, rather than local, sources. That's how I've downgraded.

Thank you very much for your help. I'm not sure how I missed it, perhaps it wasn't clear to me that `downgrade` is a separate package/command, but eventually I figured it out, and was indeed able to downgrade rather painlessly.

I'm not sure why it didn't kick in right away, but eventually the glitches went away, after a couple of reboots. Here are the versions I currently have installed which work for me:

<code>
[manjaro-virtualbox manjaro]# pacman -Q | grep mesa
lib32-libva-mesa-driver 21.2.5-1
lib32-mesa 21.2.5-1
lib32-mesa-demos 8.4.0-3
libva-mesa-driver 21.2.5-1
mesa 21.2.5-1
mesa-demos 8.4.0-6
</code>

**Comment by AckslD** [comment]
> Could someone who is affected by this (or someone who doesn't mind upgrading again) please try whether QTWEBENGINE_DISABLE_GPU_THREAD=1 changes anything (should force webengine to single-threaded opengl mode, but is QtWebEngine 5.15 only)?
> 
> Personally I can't seem to reproduce this, but I'm on AMD graphics.

@The-Compiler I just tried this and seems to still give the glitches (QtWebEngine 5.15.5 and lib32-mesa 21.3.2-1).

**Comment by rien333** [comment]
Fix is finally on its way: https://gitlab.freedesktop.org/mesa/mesa/-/issues/5731#note_1202630. Hope it lands soon.

**Comment by prosoitos** [comment]
Patch was applied to 21.3.3

Fixes the problem.

**Comment by The-Compiler** [comment]
@prosoitos can you elaborate? Don't see a patch so far being integrated either upstream or in Arch

**Comment by prosoitos** [comment]
From the link that @rien333 shared: https://gitlab.freedesktop.org/mesa/mesa/-/issues/5731#note_1202630

**Comment by The-Compiler** [comment]
Note that that was an unofficial patch, and upstream didn't like the solution. Looks like there is an [official patch](https://gitlab.freedesktop.org/pzanoni/mesa/-/commit/168f177b8a910004de768b1a89a0adea383fa6bc) now, which seems to solve the issue as well. Hoping for it to be merged and released soon...

**Comment by prosoitos** [comment]
OK. And I actually realized after a bit more testing, that the unofficial patch was helping a lot but not fully solving the problem... So I wrote here a little too soon.

Glad that another patch is coming!

**Comment by wgmayer0** [comment]
I tried upgrading and am still seeing flickering so I am guessing the official patch has not been released yet

**Comment by Aster89** [comment]
Usually I observe white flashing, as described in #6894; less frequently (but most often when waking up the system from sleep) I observe the glitches demonstrated in the opening message of the present thread; sometimes a page goes white and stays so untill I scroll on it (or maybe do another interaction that triggers a re-draw, I guess), as shown below:

![flash](https://user-images.githubusercontent.com/20521900/150101286-dee1f585-d875-433b-bf04-573b2537a611.gif)

Do you think it is in fact showing the same problem?

**Comment by The-Compiler** [comment]
@Aster89 Most likely. If you want to be sure, you might want to try downgrading mesa.

**Comment by rien333** [comment]
@Aster89 Do you use an intel UHD 6XX series video card? From what I know, the glitch only occurs on those. If you do have a card from that series, its probably the same glitch: the way it looks/manifests seemed to depend on the website and user interaction.

**Comment by Aster89** [comment]
@rien333,
```
$ lspci | grep -i nvidia
01:00.0 3D controller: NVIDIA Corporation GP108M [GeForce MX250] (rev a1)
```
So I guess I don't use a UHD 6XX series video card, am I correct?

However,
```
$ lspci | grep -i UHD
00:02.0 VGA compatible controller: Intel Corporation CometLake-U GT2 [UHD Graphics] (rev 02)
```
... mmm, I'm a bit ignorant on the topic.

**Comment by 4nd3r** [comment]
FYI Fixed in version mesa/21.3.5-1 (for Sid):
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1003987

**Comment by rien333** [comment]
Nice, 21.3.5 also landed in Arch a few days ago. I can confirm that everything seems normal so far.

 So I guess everyone can now finally remove mesa from `IgnorePkg` in `/etc/pacman.conf`.

**Comment by prosoitos** [comment]
Thanks for the update 🙂

**Comment by The-Compiler** [comment]
Looks like this has finally been fixed upstream with [mesa 21.3.6](https://docs.mesa3d.org/relnotes/21.3.6.html#bug-fixes).
